Attempting to "position" a new index breaks diff/push --verify

If an existing table has at least one secondary index, and its corresponding *.sql file is edited to add a new index such that it is *before* an existing index, this breaks `skeema diff` and `skeema push` if the `--verify` option is enabled (which it is by default):

```
~/schemas/product $ git diff
diff --git a/product/posts.sql b/product/posts.sql
index 898e6b5..07ea824 100644
--- a/product/posts.sql
+++ b/product/posts.sql
@@ -5,5 +5,6 @@ CREATE TABLE `posts` (
   `created_at` datetime DEFAULT CURRENT_TIMESTAMP,
   `edited_at` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
   PRIMARY KEY (`id`),
+  KEY `this_breaks_stuff` (`created_at`),
   KEY `user_created` (`user_id`,`created_at`)
 ) ENGINE=InnoDB DEFAULT CHARSET=latin1;

~/schemas/product $ skeema diff
2018-06-18 15:34:59 [INFO]  Generating diff of 127.0.0.1:32771 product vs /Users/evan/schemas/product/*.sql
2018-06-18 15:34:59 [ERROR] verifyDiff: Failure on table posts
DDL:
[ALTER TABLE `posts` ADD KEY `this_breaks_stuff` (`created_at`)]

EXPECTED POST-ALTER:
CREATE TABLE `posts` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `user_id` bigint(20) unsigned NOT NULL,
  `body` text,
  `created_at` datetime DEFAULT CURRENT_TIMESTAMP,
  `edited_at` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`),
  KEY `this_breaks_stuff` (`created_at`),
  KEY `user_created` (`user_id`,`created_at`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1

ACTUAL POST-ALTER:
CREATE TABLE `posts` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `user_id` bigint(20) unsigned NOT NULL,
  `body` text,
  `created_at` datetime DEFAULT CURRENT_TIMESTAMP,
  `edited_at` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`),
  KEY `user_created` (`user_id`,`created_at`),
  KEY `this_breaks_stuff` (`created_at`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1

Run command again with --skip-verify if this discrepancy is safe to ignore
```

MySQL does not offer a way to "position" new indexes, without dropping and re-creating the other indexes that are now to be "after" the new one(s). Index order in InnoDB does not have any functional difference anyway, except for in one rare edge case: in tables with no primary key but multiple unique keys with all non-nullable columns, the first unique key with non-nullable columns gets used for the clustered index.

Despite not usually having a functional impact, misordered indexes still break Skeema's validation logic, which is overly strict about ensuring it matches `SHOW CREATE TABLE` *exactly*. So the fix for this will involve:
* Provide a new option `--exact-match` which tells Skeema to respect the index order specified in the *.sql file, even if this results in purely cosmetic index drops and re-adds just to get the proper order.
* Give Skeema the ability to detect the one edge case where index order actually matters, and always respect index ordering in that case, even if `--exact-match` is not used.

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -152,7 +152,7 @@
   branch = "master"
   name = "github.com/skeema/tengo"
   packages = ["."]
-  revision = "ba3e8ce6ebae7abdaae18d7fca1fe5b56304cd04"
+  revision = "eeee978997419f1a9eb03f62fbd3dd858c3f9c2a"
 
 [[projects]]
   branch = "master"

--- a/cmd_push.go
+++ b/cmd_push.go
@@ -29,6 +29,7 @@ top of the file. If no environment name is supplied, the default is
 	cmd.AddOption(mybase.BoolOption("allow-unsafe", 0, false, "Permit running ALTER or DROP operations that are potentially destructive"))
 	cmd.AddOption(mybase.BoolOption("dry-run", 0, false, "Output DDL but don't run it; equivalent to `skeema diff`"))
 	cmd.AddOption(mybase.BoolOption("first-only", '1', false, "For dirs mapping to multiple instances or schemas, just run against the first per dir"))
+	cmd.AddOption(mybase.BoolOption("exact-match", 0, false, "Follow *.sql table definitions exactly, even for differences with no functional impact"))
 	cmd.AddOption(mybase.BoolOption("brief", 'q', false, "<overridden by diff command>").Hidden())
 	cmd.AddOption(mybase.StringOption("alter-wrapper", 'x', "", "External bin to shell out to for ALTER TABLE; see manual for template vars"))
 	cmd.AddOption(mybase.StringOption("alter-wrapper-min-size", 0, "0", "Ignore --alter-wrapper for tables smaller than this size in bytes"))
@@ -192,6 +193,7 @@ func pushWorker(sps *sharedPushState) {
 			// Set configuration-dependent statement modifiers here inside the Target
 			// loop, since the config for these may var per dir!
 			mods.AllowUnsafe = t.Dir.Config.GetBool("allow-unsafe") || sps.briefOutput
+			mods.StrictIndexOrder = t.Dir.Config.GetBool("exact-match")
 			mods.AlgorithmClause, err = t.Dir.Config.GetEnum("alter-algorithm", "INPLACE", "COPY", "DEFAULT")
 			if err != nil {
 				sps.setFatalError(NewExitValue(CodeBadConfig, err.Error()))

--- a/doc/options.md
+++ b/doc/options.md
@@ -16,6 +16,7 @@
 * [default-collation](#default-collation)
 * [dir](#dir)
 * [dry-run](#dry-run)
+* [exact-match](#exact-match)
 * [first-only](#first-only)
 * [host](#host)
 * [host-wrapper](#host-wrapper)
@@ -295,6 +296,24 @@ Commands | push
 
 Running `skeema push --dry-run` is exactly equivalent to running `skeema diff`: the DDL will be generated and printed, but not executed. The same code path is used in both cases. The *only* difference is that `skeema diff` has its own help/usage text, but otherwise the command logic is the same as `skeema push --dry-run`.
 
+### exact-match
+
+Commands | diff, push
+--- | :---
+**Default** | false
+**Type** | boolean
+**Restrictions** | none
+
+Ordinarily, `skeema diff` and `skeema push` ignore certain table differences which have no functional impact in MySQL and serve purely cosmetic purposes. For example, if two tables list their indexes in a different order, this difference is normally ignored to avoid needlessly dropping and re-adding the indexes, which may be slow if the table is large.
+
+If the [exact-match](#exact-match) option is used, these purely-cosmetic differences will be included in the generated `ALTER TABLE` statements instead of being suppressed. In other words, Skeema will attempt to make the exact table definition in MySQL exactly match the corresponding table definition specified in the *.sql file.
+
+Be aware that MySQL itself sometimes also suppresses attempts to make cosmetic changes to a table's definition! You must combine the [exact-match](#exact-match) option with [alter-algorithm=COPY](#alter-algorithm) to circumvent this behavior on the MySQL side. This forces MySQL to rebuild the table, which will be slow for large tables.
+
+Currently, [exact-match](#exact-match) only affects index ordering. In the near future, once foreign keys are supported, this option will affect foreign key naming and ordering as well.
+
+In the one case in InnoDB when index ordering has a functional impact (tables with no primary key, but multiple unique indexes over all non-nullable columns), Skeema will automatically respect index ordering, regardless of whether [exact-match](#exact-match) is enabled.
+
 ### first-only
 
 Commands | diff, push
@@ -540,4 +559,4 @@ Commands | diff, push
 
 Controls whether generated `ALTER TABLE` statements are automatically verified for correctness. If true, each generated ALTER will be tested in the temporary schema. See [the FAQ](faq.md#auto-generated-ddl-is-verified-for-correctness) for more information.
 
-It is recommended that this variable be left at its default of true, but if desired you can disable verification for speed reasons.
+It is recommended that this option be left at its default of true, but if desired you can disable verification for performance reasons.

--- a/skeema_cmd_test.go
+++ b/skeema_cmd_test.go
@@ -620,7 +620,7 @@ func (s *SkeemaIntegrationSuite) TestNonInnoClauses(t *testing.T) {
 	s.handleCommand(t, CodeSuccess, ".", "skeema push")
 	s.dbExec(t, "product", "DROP TABLE `problems`")
 	s.dbExec(t, "product", withoutClauses)
-	s.dbExec(t, "product", "ALTER TABLE `problems` DROP INDEX `idx2`")
+	s.dbExec(t, "product", "ALTER TABLE `problems` DROP KEY `idx2`")
 	writeFile(t, "mydb/product/problems.sql", withClauses)
 	s.handleCommand(t, CodeSuccess, ".", "skeema push")
 }

--- a/skeema_cmd_test.go
+++ b/skeema_cmd_test.go
@@ -296,17 +296,6 @@ func (s *SkeemaIntegrationSuite) TestDiffHandler(t *testing.T) {
 			t.Fatalf("Unable to delete diff-brief.out: %s", err)
 		}
 	}
-
-	// Confirm --verify not tripped up by unexpected index reordering
-	contents := readFile(t, "mydb/product/posts.sql")
-	lines := make([]string, 6)
-	for n := range lines {
-		lines[n] = fmt.Sprintf("KEY `idxnew_%d` (created_at)", n)
-	}
-	replacement := fmt.Sprintf("),\n  %s\n", strings.Join(lines, ",\n  "))
-	contents = strings.Replace(contents, ")\n", replacement, 1)
-	writeFile(t, "mydb/product/posts.sql", contents)
-	s.handleCommand(t, CodeDifferencesFound, ".", "skeema diff --verify")
 }
 
 func (s *SkeemaIntegrationSuite) TestPushHandler(t *testing.T) {
@@ -386,6 +375,51 @@ func (s *SkeemaIntegrationSuite) TestPushHandler(t *testing.T) {
 	s.assertMissing(t, "product", "comments", "foo")
 	s.assertMissing(t, "product", "users", "foo")
 	s.assertExists(t, "bonus", "table2", "")
+}
+
+func (s *SkeemaIntegrationSuite) TestIndexOrdering(t *testing.T) {
+	s.handleCommand(t, CodeSuccess, ".", "skeema init --dir mydb -h %s -P %d", s.d.Instance.Host, s.d.Instance.Port)
+
+	// Add 6 new redundant indexes to posts.sql. Place them before the existing
+	// secondary index.
+	contentsOrig := readFile(t, "mydb/product/posts.sql")
+	lines := make([]string, 6)
+	for n := range lines {
+		lines[n] = fmt.Sprintf("KEY `idxnew_%d` (`created_at`)", n)
+	}
+	joinedLines := strings.Join(lines, ",\n  ")
+	contentsIndexesFirst := strings.Replace(contentsOrig, "PRIMARY KEY (`id`),\n", fmt.Sprintf("PRIMARY KEY (`id`),\n  %s,\n", joinedLines), 1)
+	writeFile(t, "mydb/product/posts.sql", contentsIndexesFirst)
+
+	// push should add the indexes, and afterwards diff should report no
+	// differences, even though the index order in the file differs from what is
+	// in mysql
+	s.handleCommand(t, CodeSuccess, "", "skeema push")
+	s.handleCommand(t, CodeSuccess, "", "skeema diff")
+
+	// however, diff --exact-match can see the differences
+	s.handleCommand(t, CodeDifferencesFound, "", "skeema diff --exact-match")
+
+	// pull should re-write the file such that the indexes are now last, just like
+	// what's actually in mysql
+	s.handleCommand(t, CodeSuccess, "", "skeema pull")
+	contentsIndexesLast := strings.Replace(contentsOrig, ")\n", fmt.Sprintf("),\n  %s\n", joinedLines), 1)
+	if fileContents := readFile(t, "mydb/product/posts.sql"); fileContents == contentsIndexesFirst {
+		t.Error("Expected skeema pull to rewrite mydb/product/posts.sql to put indexes last, but file remained unchanged")
+	} else if fileContents != contentsIndexesLast {
+		t.Errorf("Expected skeema pull to rewrite mydb/product/posts.sql to put indexes last, but it did something else entirely. Contents:\n%s\nExpected:\n%s\n", fileContents, contentsIndexesLast)
+	}
+
+	// Edit posts.sql to put the new indexes first again, and ensure
+	// push --exact-match actually reorders them
+	writeFile(t, "mydb/product/posts.sql", contentsIndexesFirst)
+	s.handleCommand(t, CodeSuccess, "", "skeema push --exact-match --alter-algorithm=COPY")
+	s.handleCommand(t, CodeSuccess, "", "skeema diff")
+	s.handleCommand(t, CodeSuccess, "", "skeema diff --exact-match")
+	s.handleCommand(t, CodeSuccess, "", "skeema pull")
+	if fileContents := readFile(t, "mydb/product/posts.sql"); fileContents != contentsIndexesFirst {
+		t.Errorf("Expected skeema pull to have no effect at this point, but instead file now looks like this:\n%s", fileContents)
+	}
 }
 
 func (s *SkeemaIntegrationSuite) TestAutoInc(t *testing.T) {

--- a/target.go
+++ b/target.go
@@ -254,7 +254,10 @@ func (t *Target) verifyDiff(diff *tengo.SchemaDiff) (err error) {
 		return fmt.Errorf("verifyDiff: cannot connect to %s: %s", t.Instance, err)
 	}
 	mods := tengo.StatementModifiers{
-		NextAutoInc: tengo.NextAutoIncIgnore,
+		NextAutoInc:      tengo.NextAutoIncIgnore,
+		StrictIndexOrder: true,   // needed since we must get the SHOW CREATE TABLEs to match
+		AllowUnsafe:      true,   // needed since we're just running against the temp schema
+		AlgorithmClause:  "COPY", // needed to avoid having MySQL ignore index changes that are simply reordered
 	}
 
 	// Iterate over the ALTER-type TableDiffs in the SchemaDiff and index by table

